File: Normalising_Flows/s_volume_est.m
clear
clc
close all


syms r n rho(r) r_max

assume(n, 'integer')
assume(r_max, 'positive')

rho(r) = exp(-0.5*r^2)/sqrt((2*pi)^n);

V = int(rho(r)*r^(n-1), 0, r_max);


%% vary n

nVals = round(logspace(0,3,5));
r_max_val = 1;

figure
hold on
r_max_vals = [1,1.5,1.7];
for rIdx = 1:length(r_max_vals)
    r_max_val = r_max_vals(rIdx);
    for nIdx = 1:length(nVals)
        V_vals(nIdx) = vpa(subs(V, [n, r_max], [nVals(nIdx), r_max_val]),5);
    end
    ATplot(nVals, V_vals);
    leg(rIdx) = sprintf("r = %.2f", r_max_val);
end
set(gca, 'yscale','log', 'xscale','log')
xlabel('Dimension')
ylabel('Enclosed Probability density')
legend(leg)
ATprettify

%%

r_max_vals = linspace(0.1,2,10);
V_vals = [];

figure
hold on
n_vals = [30, 28*28];
for nIdx = 1:length(n_vals)
    n_val = n_vals(nIdx);
    for rIdx = 1:length(r_max_vals)
        V_vals(rIdx) = vpa(subs(V, [n, r_max], [n_val, r_max_vals(rIdx)]),5);
    end
    ATplot(r_max_vals, log10(V_vals));
    leg(nIdx) = sprintf("n = %d", n_val);
end
set(gca, 'xscale','log')
xlabel('r')
ylabel('log10(Enclosed probability density)')
legend(leg)
ATprettify


%% how it varies with number of images that need to be checked


nVals = round(logspace(0,3,5));
r_max_val = 1;
V_vals = [];

figure
hold on
r_max_vals = [1,1.5,1.7];
for rIdx = 1:length(r_max_vals)
    r_max_val = r_max_vals(rIdx);
    for nIdx = 1:length(nVals)
        % im part is an artefact
        V_vals(nIdx) = real(vpa(subs(V, [n, r_max], [nVals(nIdx), r_max_val]),5));
    end
    ATplot(nVals, 8*nVals*log10(2) + log10(V_vals));
    leg(rIdx) = sprintf("r = %.2f", r_max_val);
end
% set(gca, 'xscale','log')
xlabel('Dimension')
ylabel('log10(Number of images)')
legend(leg)
ATprettify


%% compare to not using NF


figure
hold on

nVals = round(logspace(0,3,5));
r_max_val = 1.5;
V_vals = [];

for nIdx = 1:length(nVals)
    % im part is an artefact
    V_vals(nIdx) = real(vpa(subs(V, [n, r_max], [nVals(nIdx), r_max_val]),5));
end
ATplot(nVals, 8*nVals*log10(2) + log10(V_vals));
ATplot(nVals, 8*nVals*log10(2))
xlabel('Dimension')
ylabel('log10(Number of images)')
legend(["With NF","Brute Force"])
ATprettify

File: Normalising_Flows/s_volume_est.py
# estimate the volume of the space of possible images

__author__ = "Adam Taras"

import sympy
import numpy as np
import matplotlib.pyplot as plt


plt.style.use('dark_background')

# define values
r = sympy.symbols('r', positive=True) # the radius in integral
r_0 = sympy.symbols('r_0', positive=True) # the max radius for n=1
r_max = sympy.symbols('r_max', positive=True) # the max radius in dim n
n = sympy.symbols('n', positive=True, integer=True) # the number of pixels

# some things to keep eqn neater
pi = sympy.pi
log10 = lambda x : sympy.log(x,10)
sqrt = sympy.sqrt
exp = sympy.exp

# the gaussian distribution
rho = sympy.exp(-0.5*r**2)/sympy.sqrt((2*sympy.pi)**n)



# consider just the integral part without any multipliers to make sure it is calculating 
int_V = sympy.integrate(rho*r**(n-1),(r,0, r_max))
int_V_fn = sympy.lambdify(n, int_V.subs(r_max, 1*sqrt(n)), 'mpmath')

nVals = np.round(np.linspace(3,1000,10))
int_V_vals = np.zeros_like(nVals)

for nIdx in range(len(nVals)):
    int_V_vals[nIdx] = float(int_V_fn(nVals[nIdx]))

plt.figure()
plt.plot(nVals, np.log10(int_V_vals))
plt.xlabel('N')
plt.ylabel('Integral part of V')
plt.show()


# the full V expression
V = log10(2*pi**(n/2)/sympy.gamma(n/2)) + log10(sympy.integrate(rho*r**(n-1), (r, 0, r_max)))

plt.figure()

# take your pick of r_0 values
r_0s = [0.6,0.8,1,1.2]
r_0s = np.linspace(1-0.05, 1+0.05,6)

for r_0val in r_0s:
    V_fn = sympy.lambdify(n, V.subs(r_max, r_0val*sqrt(n)), 'mpmath')
    
    
    nVals = np.round(np.linspace(3,10000,50))
    V_vals = np.zeros_like(nVals)
    
    for nIdx in range(len(nVals)):
        V_vals[nIdx] = float(V_fn(nVals[nIdx]))
    
    plt.plot(nVals,V_vals, label=f'r_0 = {r_0val}')
    
plt.xlabel('Dimension N')
plt.ylabel('log10(V)')
plt.legend()
plt.savefig('Vwide.pdf')
plt.ylim([-1,0])
plt.savefig('Vzoomed.pdf')
plt.show()



# now number of images

plt.figure()
r_0s = np.linspace(1-0.5, 1+0.05,6)
plt.plot(nVals,8*nVals*np.log10(2), label=f'Brute Force')

for r_0val in r_0s:
    V_fn = sympy.lambdify(n, V.subs(r_max, r_0val*sqrt(n)), 'mpmath')
    
    
    nVals = np.round(np.linspace(3,10000,50))
    V_vals = np.zeros_like(nVals)
    
    for nIdx in range(len(nVals)):
        V_vals[nIdx] = float(V_fn(nVals[nIdx]))
        
    plt.plot(nVals,8*nVals*np.log10(2) + V_vals, label=f'r_0 = {r_0val}')

plt.xlabel('Dimension N')
plt.ylabel('log10(Number of images)')

plt.legend()
plt.savefig('nImgs.svg')


